torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+    edge_list& rootEdges,
+    variable_list& grads) {
+  TORCH_CHECK(!roots.empty(), "No tensors provided for gradient computation.");
+  TORCH_INTERNAL_ASSERT(rootEdges.empty());
+  TORCH_INTERNAL_ASSERT(grads.empty());
+
+  // Verify roots are all scalar and require gradients.
+  for (const auto& root : roots) {
+    TORCH_CHECK(
+        root.requires_grad(), "requires_grad not set on: ", root.name());
+    TORCH_CHECK(
+        root.numel() == 1,
+        root.name(),
+        " is not a scalar, all roots need to be scalar");
+    TORCH_CHECK(
+        root.grad_fn(),
+        root.name(),
+        " does not have a valid gradient function.");
+
+    // Compute the root edges and generate the appropriate gradients.
+    rootEdges.push_back(root.gradient_edge());
+    grads.push_back(at::ones_like(root));
+  }
+
+  // Validate rootEdges and grads.
+  validate_outputs(
+      rootEdges, grads, [](const std::string& msg) { return msg; });
+}
+
+void DistEngine::computeDependencies(
+    DistAutogradContext& autogradContext,
+    const edge_list& rootEdges,
+    const variable_list& grads,
+    std::shared_ptr<Node>& graphRoot,
+    edge_list& outputEdges) {
+  // Build the graph task and graph root.
+  auto graphTask = std::unique_ptr<GraphTask>(new GraphTask(
+      /* keep_graph */ false, /* create_graph */ false, /* depth */ 0));
+
+  if (graphRoot == nullptr) {
+    // Create a graphRoot only if needed.
+    graphRoot = std::make_shared<GraphRoot>(rootEdges, grads);
+  }
+
+  // Run BFS to traverse the graph locally. The roots of the graph are
+  // GraphRoot and all send functions for this autograd context.
+  std::unordered_set<Node*> seen;
+  std::queue<Node*> queue;
+  queue.push(static_cast<Node*>(graphRoot.get()));
+
+  auto sendFunctions = autogradContext.sendFunctions();
+
+  // Add all the send functions to the queue as roots.
+  for (const auto& mapEntry : sendFunctions) {
+    // Make sure we don't enqueue the same function twice.
+    if (mapEntry.second != graphRoot) {
+      // Increment 'outstanding_tasks_' for GraphTask for each send_function
+      // since we want the local autograd engine to wait for all of them.
+      graphTask->outstanding_tasks_++;
+      queue.push(mapEntry.second.get());
+    }
+  }
+
+  edge_list recvBackwardEdges;
+  // Traverse the graph.
+  auto& dependencies = graphTask->dependencies_;
+  while (!queue.empty()) {
+    auto fn = queue.front();
+    queue.pop();
+
+    for (const auto& edge : fn->next_edges()) {
+      if (auto nextFn = edge.function.get()) {
+        dependencies[nextFn] += 1;
+        const bool wasInserted = seen.insert(nextFn).second;
+        if (wasInserted) {
+          // Seeing this function for the first time.
+          queue.push(nextFn);
+
+          if (nextFn->next_edges().empty()) {
+            TORCH_INTERNAL_ASSERT(
+                dynamic_cast<AccumulateGrad*>(nextFn) ||
+                dynamic_cast<RecvRpcBackward*>(nextFn));
```

Comment:
RTTI! (I guess it's OK here because we're unlikely to put this code on mobile)

Reply:
I'm not too happy about this either, is there a better way to do this? I couldn't find a better way of figuring out what type of node we're looking at.

Reply:
You'd have to add tagging of some sort. Maybe a virtual method that says whether or not this is a "leaf" node or not (default to false, AccumulateGrad and RecvRpcBackward make it true.)

Reply:
In other places we also do need to check if this is AccumulateGrad (and not RecvRpcBackward). One idea I had was having a `get_custom_name` API in `Node`. This way AccumulateGrad and RecvRpcBackward can set this and we can retrieve the name to check if we're looking at the same node. Instead of names (strings), we could also just have an enum and have a method like `get_node_type`. For now the enum could have only three values { ACCUMULATE_GRAD, RECV_RPC_BACKWARD, UNKNOWN}; Although, I'd like to do this in a separate PR.